src/CodeRunner.ts

```diff
@@ -119,9 +119,8 @@ export class CodeRunner {
             type: BackendEventType.End,
             data: "User cancelled run", contentType: "text/plain"
         });
-        // Since we use workers, the old one must be entirely replaced to interrupt it
-        this.backend.interrupt();
-        return this.start();
+        await this.backend.interrupt();
+        this.setState(RunState.Ready);
```

Comment:
Interrupting no longer necessarily means that the program will stop, because KeyboardInterrupt can be caught. pdb in particular does that. You need to catch `InterruptError` where `client.call` is called, and that will indicate that the worker was terminated because other interruption methods weren't available.

Reply:
By default, calling interrupt on a PyodideClient leads to a PythonError insteads of an InterruptError as the internal KeyBoardError is raised within Pyodide. Is your expected way of handling this to catch this in run_async?  Don't we want to throw an InterruptError in SyncClient.call by default, or should I use a custom interrupter for this?

Reply:
The user's code might catch `KeyboardInterrupt` (a notable case is pdb) and there's no way for the client to know whether that will happen. It mustn't throw InterruptError because the worker may not be done processing and ready for more.

Letting `KeyboardInterrupt` bubble up from python_runner's _execute_context without going through serialize_traceback is useful in futurecoder where significant work usually happens after running code to check if the answer was correct for the current step. But for the sake of generality, I suggest removing that case from _execute_context, and futurecoder can add a bare `raise` in serialize_traceback.